### ISSUE TITLE
Query/Aggregate Builders no longer allow incremental results / always cache

### ISSUE THREAD

**Issue by Steveb-p** [doctrine/mongodb-odm#2113]
|    Q        |   A
|------------ | -----
| Version     | 2.0.2

### Support Question

In ODM 1.3 I was able to build and run queries with large result sets. This allowed me to process incoming data incrementally, by yielding data in generator.

In 2.0 with new Builders I've noticed that both builders wrap `Cursor` with `CachingIterator`, which causes high memory usage in such cases. I've hit my memory limit (obviously didn't want to raise it, I had it relatively high in the first place) and had to work around it - I'm getting pipeline built and passed directly into `aggregate` method of `Collection`.

```php
$builder = $this->dm->createAggregationBuilder($collectionName);

$collection = $this->dm->getClient()->selectCollection('xxx', $collectionName);
yield from $collection->aggregate($builder->getPipeline(), [
    'allowDiskUse' => true,
    'cursor' => true,
]);
``` 

I'd like to be able to select which class caches the results, maybe discarding the ability to rewind/count when I know I won't be using them?

**Comment by alcaeus** [comment]
Ah, I see. We originally made the decision to wrap the MongoDB cursor in a `CachingIterator` instance to allow rewinding over the result set. I think adding an explicit way to disable the `CachingIterator` functionality should be possible for query and aggregation builders:

```php
$builder
    ->cache(false) // true is the default
    ->getQuery()
    ->getIterator(); // Would no longer return a `CachingIterator`
```

For the aggregation builder I'd have to think about where to put this option. I'm a bit vary of putting it on the root level (like we do in query builder) since we only have aggregation pipeline stages there. On the other hand, aggregation options in `execute` might be wrong as well, as those options are all passed to the underlying `aggregate` method.

If you want to give implementing this a shot, please feel free to reach out on Slack or here to discuss specifics.

**Comment by malarzm** [comment]
FWIW I'd name the method `rewindable` or something alike. Cache might imply we finally implemented 2nd level cache for queries :)

**Comment by alcaeus** [comment]
Good point. Let's not call it `cache` ;)

**Comment by Steveb-p** [comment]
> If you want to give implementing this a shot, please feel free to reach out on Slack or here to discuss specifics.

Gladly, however I've still got another issue I've reported here that I promised to take care of, so I don't want to promise more :) I'll try though. I'll say so here if I start doing anything about it :smile:

**Comment by Steveb-p** [comment]
I've looked into both `HydratingIterator` and `CachingIterator` and their behavior is a little different. By this I mean `HydratingIterator` does not cache it's results and does not allow re-iterating over the results. I've confirmed this by adding a test to `HydratingIteratorTest`:

```php
public function testRewindAfterPartialIteration()
{
    $iterator = new HydratingIterator($this->getTraversable([1, 2, 3]), $this->uow, $this->dm->getClassMetadata(User::class));

    $iterator->rewind();
    $this->assertTrue($iterator->valid());

    $iterator->next();
    $this->expectException(Throwable::class);
    $this->expectExceptionMessage('Cannot rewind a generator that was already run');
    iterator_to_array($iterator);
}
```
which passes. This is in contrast to a similar test that is present for `CachingIterator`:
https://github.com/doctrine/mongodb-odm/blob/3b8241869d8540c91eb7a604575ea09205ecc5fd/tests/Doctrine/ODM/MongoDB/Tests/Functional/Iterator/CachingIteratorTest.php#L14-L25

I believe that if we are to allow control over whether the returned iterator caches his results or not should be consistent between these two.

Therefore, I'm wondering if it wouldn't be "better" to make the iterator remember the original `Query` object instead and re-run query on rewind, leaving eventual wrapping with caching iterator to end-user? What was the reasoning behind introducing this internal cache? What is it used for?

**Comment by alcaeus** [comment]
> I've looked into both `HydratingIterator` and `CachingIterator` and their behavior is a little different. By this I mean `HydratingIterator` does not cache it's results and does not allow re-iterating over the results. [...]
> Therefore, I'm wondering if it wouldn't be "better" to make the iterator remember the original `Query` object instead and re-run query on rewind, leaving eventual wrapping with caching iterator to end-user? What was the reasoning behind introducing this internal cache? What is it used for?

This is intentional. For a bit of context, let me elaborate on cursor in MongoDB. Server-side cursors are forward-only: once you have received a batch, there is no going back to the previous batch. Once the cursor is exhausted, i.e. the last batch of data has been sent, the cursor is closed on the server. In the legacy `ext-mongo` extension, this was hidden away in the `MongoCursor` class. When you called `rewind` on the cursor, it ran the command. This means that if you iterate a `MongoCursor` twice, the second iteration might get you different results. Similarly, calling `count` on a `MongoCursor` didn't count the number of results, but rather sent a `count` command to the server. Again, this could return a different number of results than what you would see by iterating on the cursor.

When creating the current `ext-mongodb` extension, this behaviour was changed to be consistent to the server. The `MongoDB\Driver\Cursor` class is forward-only and cannot be rewound or counted. This was done to remove the issues I mentioned in the paragraph above.

When we look at this from an ODM context, we expect that in some cases, users may want to iterate cursors multiple times or count the number of results returned from the cursor. To allow for this, we added the `CachingIterator` class that iterates the underlying cursor during the first iteration, then iterates the cached results on the subsequent iteration. When running `count` on such a cursor or when rewinding a partially iterated cursor, the underlying cursor object is exhausted and all results stored in memory.

The reason the `HydratingIterator` does not show this behaviour is because we don't want to run hydration multiple times when rewinding the iterator. Thus, we cache hydrated results for performance reasons, at the expense of a higher memory footprint. This could technically be changed by wrapping the underlying cursor in a `CachingIterator`, then wrapping that in a `HydratingIterator` to save a little bit of memory at the expense of higher CPU usage due to the additional hydration calls.

I understand your use-case, and the goal is to create a feature to allow users more control over how their data is cached. One option would be to indeed create a cursor class that stores the original query/aggregation object and re-runs this when the cursor is rewound. The tricky part is making this behaviour explicit and understandable for users. This includes properly documenting the API as well as the performance implications of disabling caching on the cursor. However, I'm not sure if this is an API we want to introduce here, but I'll be discussing this with @malarzm and @jmikola at SymfonyCon later this week.

**Comment by Steveb-p** [comment]
@alcaeus Thanks for the detailed explanation :+1: It perfectly answers my doubts regarding the reasoning of the change between previous major version.

I've read the usage of `HydratingIterator` wrong. I thought that it is returned immediately. I did not notice that practically always it is wrapped with `CachingIterator`. I expected that the memory issue would occur only when not hydrating results, which is not the case.

I would say then - if I can voice my opinion on this matter - that it is reasonable to add `rewindable` method (and option) to normal and aggregation queries (builders) and return iterators without wrapping them in `CachingIterator` when it is true. Otherwise it is almost impossible to iterate over large result sets, with hydration or not.

**Comment by sirian** [comment]
+1 to this issue. As workaround I've implemented helper function, which unwrap caching iterator via reflection. 

```php
class DoctrineUtil
{
    public static function iterate($query)
    {
        if ($query instanceof \Doctrine\ODM\MongoDB\Query\Builder) {
            $query = $query->getQuery();
        }
        if ($query instanceof \Doctrine\ODM\MongoDB\Aggregation\Builder) {
            $query = $query->getQuery();
        }
        if (!$query instanceof \Doctrine\ODM\MongoDB\Query\Query) {
            throw new \InvalidArgumentException(sprintf('Expected Query or Builder, given "%s"', get_class($query)));
        }

        $iterator = $query->execute();

        if ($iterator instanceof \Doctrine\ODM\MongoDB\Iterator\CachingIterator) {
            $ref = new \ReflectionObject($iterator);
            $itemsProp = $ref->getProperty('items');
            $itemsProp->setAccessible(true);

            yield from $itemsProp->getValue($iterator);

            $iteratorProp = $ref->getProperty('iterator');
            $iteratorProp->setAccessible(true);
            yield from $iteratorProp->getValue($iterator);
        }

        return $iterator;
    }
}
```

**Comment by alcaeus** [comment]
Fixed in #2116.
